### PR TITLE
fix: CLI filtering --spec when commas in glob pattern

### LINKED ISSUE #16102: CLI filtering `--spec` with glob pattern does not find spec files
### Current behavior
Given the following folder structure:
```
.
├── cypress
│   ├── integration
│       └── _fixtures
│           └── fixture-should-not-match.spec.js
│       └── _unit
│           └── unit-should-not-match.spec.js                  
│       └── sub-level-1
│           └── should-match-level-1.spec.js                   --> this one should match!
│         └── sub-level-2
│             └── should-match-level-2.spec.js                 --> this one should match!
│       └── fixture-should-not-match.spec.js
```
I'd like to trigger cypress with `cypress run` targeting only the two spec files highlighted. One glob pattern tested from [here](https://www.digitalocean.com/community/tools/glob?comments=true&glob=cypress%2Fintegration%2F%7B%5B%21fix%5D%2A.spec.js%2C%21%28_unit%29%2F%2A%2A%2F%5B%21fix%5D%2A.spec.js%7D&matches=false&tests=cypress%2Fintegration%2F_fixtures%2Ffixture-should-not-match.spec.js&tests=cypress%2Fintegration%2F_unit%2Funit-should-not-match.spec.js&tests=cypress%2Fintegration%2Fsub-level-1%2Fshould-match-level-1.spec.js&tests=cypress%2Fintegration%2Fsub-level-2%2Fshould-match-level-2.spec.js&tests=cypress%2Fintegration%2Ffixture-should-not-match.spec.js) gives the right result.

However, when triggering from the command line with `npx cypress run --spec 'cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js}'`, I get the message 
```
Can't run because no spec files were found.

We searched for any files matching this glob pattern:

cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js}

Relative to the project root folder:

/Users/username/path/to/project

```

### Desired behavior

Cypress should find & run the two specs highlighted in the snippet above.

### Test code to reproduce

`npx cypress run --spec 'cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js}'`

### Versions
Cypress 6.5.0 (Upgrade to Cypress 7 impossible atm, as the project has tied dependencies)
Mac OS 10.15.7

### PR DESCRIPTION
**Pull request by hang-up**
- Closes https://github.com/cypress-io/cypress/issues/16102

### User facing changelog
Enable comma separated glob pattern matching for CLI parameter `spec`. Fixes #16102

### Additional details
This PR aims at fixing a bug discovered since `v6.5.0`.
When glob patterns were defined with at least one comma, the original implementation would assume that when parsed as a string, the `spec` argv could be split into an array. 

This assumption holds when the argv is a string of spec files to run: 
- `--spec "cypress/integration/spec-1.spec.js,cypress/integration/spec-2.spec.js"` would give an array similar to
```
[
  "cypress/integration/spec-1.spec.js", 
  "cypress/integration/spec-2.spec.js"
]
```

Unfortunately in the case of comma separated glob pattern, we would end with erroneous paths:
- `--spec 'cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.specs.js}'` would give an array similar to
```
[
  "cypress/integration/{[!fix]*.spec.js", 
  "!(_unit)/**/[!fix]*.specs.js"
]
```

This PR fixes this issue by performing a glob check before mapping the string to an array of spec paths.

### How has the user experience changed?

Following the same file structure as defined in the issue ticket
```
.
├── cypress
│   ├── integration
│       └── _fixtures
│           └── fixture-should-not-match.spec.js
│       └── _unit
│           └── unit-should-not-match.spec.js                  
│       └── sub-level-1
│           └── should-match-level-1.spec.js                   --> this one should match!
│         └── sub-level-2
│             └── should-match-level-2.spec.js                 --> this one should match!
│       └── fixture-should-not-match.spec.js
```
**Before**
image

**After**
image

### PR Tasks

- [ ] Have tests been added/updated?
- [na] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [na] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [na] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
- [na] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by hang-up** [comment]
I'm just realizing now that some snapshot tests are failing.
I'll await the initial feedback on this PR before taking any steps further (eg. update those snapshots).

**Review by jennifer-shehane — CHANGES_REQUESTED** [review]
@hang-up The change has the unintended side effect of updating the value that is printed to CLI when showing what Cypress searched for in the specs. We'd like to maintain the original behavior or printing the original search value and not the found values (since this is already printed in the Specs heading). 

Can you update the code to not change the value?

Screen Shot 2022-01-05 at 1 02 15 PM

**Comment by hang-up** [comment]
@jennifer-shehane The side effect should have been addressed :)

**Comment by mjhenkes** [comment]
I believe this solution will break the case where we have multiple globs or a mixture of globs and specific spec files.

ie.
```
--spec"cypress/integration/examples/actions.spec.js,cypress/integration/examples/files.spec.js,cypress/integration/login/**/*"
```
indicating that you want:
* cypress/integration/examples/actions.spec.js
* cypress/integration/examples/files.spec.js
* and any specs that match cypress/integration/login/**/*

I think a better solution would involve escaping the comma and more complex parsing of the string than our current [comma split](https://github.com/cypress-io/cypress/blob/develop/packages/server/lib/util/args.js#L121).

**Comment by hang-up** [comment]
Hey @mjhenkes!
Thanks for the feedback. I've updated my array of test files/folders to the following:

```
.
├── cypress
│   ├── integration
│       └── _fixtures
│           └── fixture-should-match.spec.js        -> should match
│       └── _unit
│           └── unit-should-match.spec.js           -> should match     
│       └── login
│           └── login.spec.js                       -> should match
│       └── sub-level-1
│           └── should-not-match-level-1.spec.js                   
│         └── sub-level-2
│             └── should-not-match-level-2.spec.js                 
│       └── fixture-should-not-match.spec.js
```

And running a mixture of 2 specs + 1 glob pattern:
```
yarn cypress:run -- --spec='cypress/integration/login/login.spec.js,cypress/integration/_unit/unit-should-match.spec.js,cypress/integration/_fixtures/**/*.spec.js'
```
seems to pick up the right specs to run.

image

**Comment by mjhenkes** [comment]
I dug into the code some more, The issue will arise when there is a square or curly bracket in t regular expression. If found the whole `--spec` string will be treated as a single glob.

I believe the following string will fail for you (I have a different folder setup for my test repo so i can't be sure 😄 )
```
yarn cypress:run -- --spec='cypress/integration/login/login.spec.js,cypress/integration/_unit/unit-should-match.spec.js,cypress/integration/**/{login,fixture-should-match}.spec.js'
```

**Comment by hang-up** [comment]
You're right! 

I've narrowed it down to these 8 possible scenarios (to make sure I'm not missing anything):
```js
const patterns = {
  0: "cypress/integration/glob/**/*",
  1: "cypress/integration/spec1.spec.js",
  2: "cypress/integration/spec1.spec.js,cypress/integration/spec2.spec.js",
  3: "cypress/integration/spec1.spec.js,cypress/integration/spec2.js,cypress/integration/glob/**/*",  
  4: "{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js}",
  5: "cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js},cypress/integration/spec1.spec.js",
  6: "cypress/{static,build/public,{src,src/increment}}/*.js",
  7: "cypress/integration/{[!fix]*.spec.js,!(_unit)/**/[!fix]*.spec.js},cypress/integration/spec1.spec.js,cypress/integration/{[!fix]*.specs.js,!(_unit)/**/[!fix]*.specs.js}"
};
```
I'll do a rewrite of the split logic to accommodate them :)

File: packages/server/lib/util/args.js [hunk 1]
```diff
@@ -204,6 +204,112 @@ const sanitizeAndConvertNestedArgs = (str, argname) => {
   }
 }
 
+const parseSpecArgv = (pattern) => {
```
Comment (mjhenkes) [inline packages/server/lib/util/args.js:207]: Could you document this function like you have the inner functions?

**Comment by mjhenkes** [comment]
@hang-up your spec test cases look pretty good, I don't think i have anything to add.

File: packages/server/lib/util/args.js [hunk 2]
```diff
@@ -204,6 +204,112 @@ const sanitizeAndConvertNestedArgs = (str, argname) => {
...
+  const sliceAndMutate = (str, end) => {
+    return [
+      str.slice(0, end),
+      str.substring(end, str.length),
+      str.slice(0, end).length,
+    ]
+  }
+
+  /**
+   * Sanitizes a path from leftover commas.
```
[the remark this answers is not in the document]
Comment (hang-up) [reply packages/server/lib/util/args.js]: Done!

**Review by mjhenkes — DISMISSED** [review]

**Review by davidmunechika — APPROVED** [review]

File: packages/server/lib/util/args.js [hunk 3]
```diff
@@ -204,6 +204,118 @@ const sanitizeAndConvertNestedArgs = (str, argname) => {
...
+    OPEN: ['{', '['],
+    CLOSE: ['}', ']'],
+  }
+  const hasToken = [...TOKENS.OPEN, ...TOKENS.CLOSE].some((t) => {
+    return pattern.includes(t)
+  })
+  const hasComma = pattern.includes(',')
+
+  /**
+   * Slice & mutate a string.
```
Comment (mjhenkes) [inline packages/server/lib/util/args.js:224]:
```suggestion
   * Slice and mutate a string.
```

**Review by mjhenkes — APPROVED** [review]

**Review by davidmunechika — APPROVED** [review]
lgtm

**Pull request merged**